From 28974436a2b431f5c80bb8b3ef40e6a2654e7f58 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Wed, 19 Jun 2024 18:21:46 +0200
Subject: [PATCH] mailbox: add tx done by ack to all device having and fix mbox
 free

Devices having irq capability have the property tx_done_method set to
TXDONE_BY_IRQ, but they can support TXDONE_BY_ACK. Reinitialize tx_done method from channel at free channel.

Change-Id: I19d3ec92cf9d2048120566e579a0cd498a7b5a37
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/402234
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 drivers/mailbox/mailbox.c | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

--- a/drivers/mailbox/mailbox.c
+++ b/drivers/mailbox/mailbox.c
@@ -336,7 +336,7 @@ static int __mbox_bind_client(struct mbo
 	chan->cl = cl;
 	init_completion(&chan->tx_complete);
 
-	if (chan->txdone_method	== TXDONE_BY_POLL && cl->knows_txdone)
+	if (cl->knows_txdone)
 		chan->txdone_method = TXDONE_BY_ACK;
 
 	spin_unlock_irqrestore(&chan->lock, flags);
@@ -477,6 +477,16 @@ struct mbox_chan *mbox_request_channel_b
 }
 EXPORT_SYMBOL_GPL(mbox_request_channel_byname);
 
+static void  mbox_txdone_init(struct mbox_controller *mbox, int *tx_method)
+{
+	if (mbox->txdone_irq)
+		*tx_method = TXDONE_BY_IRQ;
+	else if (mbox->txdone_poll)
+		*tx_method = TXDONE_BY_POLL;
+	else /* It has to be ACK then */
+		*tx_method = TXDONE_BY_ACK;
+}
+
 /**
  * mbox_free_channel - The client relinquishes control of a mailbox
  *			channel by this call.
@@ -496,8 +506,8 @@ void mbox_free_channel(struct mbox_chan
 	spin_lock_irqsave(&chan->lock, flags);
 	chan->cl = NULL;
 	chan->active_req = NULL;
-	if (chan->txdone_method == TXDONE_BY_ACK)
-		chan->txdone_method = TXDONE_BY_POLL;
+
+	mbox_txdone_init(chan->mbox, &chan->txdone_method);
 
 	module_put(chan->mbox->dev->driver->owner);
 	spin_unlock_irqrestore(&chan->lock, flags);
@@ -530,12 +540,7 @@ int mbox_controller_register(struct mbox
 	if (!mbox || !mbox->dev || !mbox->ops || !mbox->num_chans)
 		return -EINVAL;
 
-	if (mbox->txdone_irq)
-		txdone = TXDONE_BY_IRQ;
-	else if (mbox->txdone_poll)
-		txdone = TXDONE_BY_POLL;
-	else /* It has to be ACK then */
-		txdone = TXDONE_BY_ACK;
+	mbox_txdone_init(mbox, &txdone);
 
 	if (txdone == TXDONE_BY_POLL) {
 
